File: VideoGamesBacklogBackend/Services/CommunityService.cs
using Microsoft.EntityFrameworkCore;
using VideoGamesBacklogBackend.Data;
using VideoGamesBacklogBackend.Dto;
using VideoGamesBacklogBackend.Interfaces;
using VideoGamesBacklogBackend.Helpers;
using AutoMapper;
using VideoGamesBacklogBackend.Entities;

namespace VideoGamesBacklogBackend.Services
{
    public class CommunityService(
        AppDbContext context,
        ILogger<CommunityService> logger,
        INotificationService notificationService,
        IMapper mapper)
        : ICommunityService
    {
        public async Task<CommunityStatsDto> GetCommunityStatsAsync(string gameTitle)
        {
            var allGames = await context.Games
                .Select(g => new
                {
                    g.Id,
                    g.Title,
                    g.HoursPlayed,
                    g.Status,
                    g.Rating,
                    HasPublicReview = g.Review != null && g.Review.IsPublic == true
                })
                .ToListAsync();

            var games = allGames.Where(g => GameTitleMatcher.DoesGameTitleMatch(g.Title, gameTitle)).ToList();
            if (games.Count == 0)
            {
                return new CommunityStatsDto();
            }

            var totalPlayers = games.Count;
            var gamesWithPublicReviews = games.Where(g => g.HasPublicReview).ToList();
            var totalReviews = gamesWithPublicReviews.Count;
            var gamesWithRating = games.Where(g => g.Rating > 0).ToList();
            var averageRating = gamesWithRating.Count > 0
                ? gamesWithRating.Average(g => g.Rating)
                : 0;

            var averagePlaytime = games.Count > 0
                ? (int)Math.Round(games.Average(g => g.HoursPlayed))
                : 0;

            var completedGames = games.Count(g => g.Status is GameStatus.Completed or GameStatus.Platinum);
            var completionRate = totalPlayers > 0
                ? Math.Round((decimal)completedGames / totalPlayers * 100, 2)
                : 0;

            var currentlyPlaying = games.Count(g => g.Status == GameStatus.InProgress);

            return new CommunityStatsDto
            {
                TotalPlayers = totalPlayers,
                AverageRating = Math.Round(averageRating, 2),
                TotalReviews = totalReviews,
                AveragePlaytime = averagePlaytime,
                CompletionRate = completionRate,
                CurrentlyPlaying = currentlyPlaying
            };
        }

        public async Task<decimal> GetCommunityRatingAsync(string gameTitle)
        {
            var games = await context.Games
                .Where(g => g.Rating > 0)
                .ToListAsync();

            games = games.Where(g => GameTitleMatcher.DoesGameTitleMatch(g.Title, gameTitle)).ToList();

            return games.Count > 0 ? Math.Round(games.Average(g => g.Rating), 2) : 0;
        }

        public async Task<CommunityRatingDto> GetCommunityRatingWithCountAsync(string gameTitle)
        {
            var games = await context.Games
                .Where(g => g.Rating > 0)
                .ToListAsync();

            games = games.Where(g => GameTitleMatcher.DoesGameTitleMatch(g.Title, gameTitle)).ToList();

            var reviewCount = games.Count;
            var rating = reviewCount > 0 ? Math.Round(games.Average(g => g.Rating), 2) : 0;

            return new CommunityRatingDto
            {
                Rating = rating,
                ReviewCount = reviewCount
            };
        }

        public async Task<Dictionary<string, decimal>> GetCommunityRatingsAsync(List<string> gameTitles)
        {
            var ratings = new Dictionary<string, decimal>();

            foreach (var title in gameTitles)
            {
                var rating = await GetCommunityRatingAsync(title);
                ratings[title] = rating;
            }

            return ratings;
        }

        public async Task<Dictionary<string, CommunityRatingDto>> GetCommunityRatingsWithCountAsync(
            List<string> gameTitles)
        {
            var ratingsWithCount = new Dictionary<string, CommunityRatingDto>();

            foreach (var title in gameTitles)
            {
                var ratingData = await GetCommunityRatingWithCountAsync(title);
                ratingsWithCount[title] = ratingData;
            }

            return ratingsWithCount;
        }

        public async Task<PaginatedReviewsDto> GetReviewsAsync(string gameTitle, int page, int pageSize,
            int? currentUserId = null)
        {
            var allGames = await context.Games
                .Include(g => g.User)
                .Include(g => g.ReviewComments)
                .Where(g => g.Review != null && g.Review.IsPublic == true)
                .ToListAsync();

            var games = allGames.Where(g => GameTitleMatcher.DoesGameTitleMatch(g.Title, gameTitle))
                .Where(g => !currentUserId.HasValue || g.UserId != currentUserId.Value)
                .OrderByDescending(g => g.Review!.Date)
                .ToList();

            var totalCount = games.Count;
            var totalPages = (int)Math.Ceiling((double)totalCount / pageSize);
            var paginatedGames = games
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .ToList();
            var reviews = paginatedGames.Select(mapper.Map<CommunityReviewDto>).ToList();

            return new PaginatedReviewsDto
            {
                Reviews = reviews,
                TotalCount = totalCount,
                Page = page,
                PageSize = pageSize,
                TotalPages = totalPages
            };
        }

        public async Task<PaginatedReviewsDto> GetPublicReviewsAsync(string gameTitle, int page, int pageSize)
        {
            var allGames = await context.Games
                .Include(g => g.User)
                .Include(g => g.ReviewComments)
                .Where(g => g.Review != null && g.Review.IsPublic == true)
                .ToListAsync();

            var games = allGames.Where(g => GameTitleMatcher.DoesGameTitleMatch(g.Title, gameTitle))
                .OrderByDescending(g => g.Review!.Date)
                .ToList();

            var totalCount = games.Count;
            var totalPages = (int)Math.Ceiling((double)totalCount / pageSize);
            var paginatedGames = games
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .ToList();
            var reviews = paginatedGames.Select(mapper.Map<CommunityReviewDto>).ToList();

            return new PaginatedReviewsDto
            {
                Reviews = reviews,
                TotalCount = totalCount,
                Page = page,
                PageSize = pageSize,
                TotalPages = totalPages
            };
        }

        public async Task<ReviewStatsDto> GetReviewStatsAsync(string gameTitle)
        {
            var allGames = await context.Games
                .Where(g => g.Review != null && g.Review.IsPublic == true)
                .ToListAsync();

            var games = allGames.Where(g => GameTitleMatcher.DoesGameTitleMatch(g.Title, gameTitle)).ToList();

            if (!(games.Count > 0))
            {
                return new ReviewStatsDto { GameTitle = gameTitle };
            }

            var reviews = games.Select(g => g.Review!).ToList();
            var totalReviews = reviews.Count;

            return new ReviewStatsDto
            {
                GameTitle = gameTitle,
                TotalReviews = totalReviews,
                AverageGameplay = Math.Round(reviews.Average(r => r.Gameplay), 2),
                AverageGraphics = Math.Round(reviews.Average(r => r.Graphics), 2),
                AverageStory = Math.Round(reviews.Average(r => r.Story), 2),
                AverageSound = Math.Round(reviews.Average(r => r.Sound), 2),
                OverallAverageRating = Math.Round(games.Average(g => g.Rating), 2),
                RatingDistribution = games
                    .GroupBy(g => (int)Math.Round(g.Rating, MidpointRounding.AwayFromZero))
                    .ToDictionary(g => g.Key, g => g.Count()),
                GameplayStats = CalculateAspectStats(reviews.Select(r => r.Gameplay)),
                GraphicsStats = CalculateAspectStats(reviews.Select(r => r.Graphics)),
                StoryStats = CalculateAspectStats(reviews.Select(r => r.Story)),
                SoundStats = CalculateAspectStats(reviews.Select(r => r.Sound))
            };
        }

        public async Task<List<CommunityReviewDto>> GetTopReviewsAsync(string gameTitle, int limit,
            int? currentUserId = null)
        {
            var allGames = await context.Games
                .Include(g => g.User)
                .Include(g => g.ReviewComments)
                .Where(g => g.Review != null && g.Review.IsPublic == true)
                .ToListAsync();

            var games = allGames.Where(g => GameTitleMatcher.DoesGameTitleMatch(g.Title, gameTitle))
                .Where(g => !currentUserId.HasValue || g.UserId != currentUserId.Value)
                .OrderByDescending(g => g.Rating)
                .ThenByDescending(g => g.Review!.Date)
                .Take(limit)
                .ToList();
            return games.Select(mapper.Map<CommunityReviewDto>).ToList();
        }

        /// <summary>
        /// Ottiene i commenti per una specifica recensione
        /// </summary>
        public async Task<List<ReviewCommentDto>> GetReviewCommentsAsync(int reviewGameId)
        {
            var comments = await context.ReviewComments
                .Include(rc => rc.Author)
                .Where(rc => rc.ReviewGameId == reviewGameId)
                .OrderBy(rc => rc.Date)
                .ToListAsync();

            return mapper.Map<List<ReviewCommentDto>>(comments);
        }

        /// <summary>
        /// Aggiunge un commento a una recensione
        /// </summary>
        public async Task<ReviewCommentDto?> AddReviewCommentAsync(CreateReviewCommentDto createCommentDto,
            int authorId)
        {
            var reviewGame = await context.Games
                .Include(g => g.User)
                .FirstOrDefaultAsync(g => g.Id == createCommentDto.ReviewGameId
                                          && g.Review != null
                                          && g.Review.IsPublic == true);

            if (reviewGame == null)
                throw new KeyNotFoundException("Recensione non trovata o non pubblica.");

            var newComment = new ReviewComment
            {
                Text = createCommentDto.Text,
                Date = DateTime.UtcNow.ToString("yyyy-MM-dd"),
                AuthorId = authorId,
                ReviewGameId = createCommentDto.ReviewGameId
            };

            context.ReviewComments.Add(newComment);
            await context.SaveChangesAsync();

            var savedComment = await context.ReviewComments
                .Include(rc => rc.Author)
                .FirstOrDefaultAsync(rc => rc.Id == newComment.Id);

            if (savedComment == null) return null;

            try
            {
                await notificationService.CreateReviewCommentNotificationAsync(
                    reviewGame.UserId,
                    authorId,
                    savedComment.Author?.UserName ?? "Utente sconosciuto",
                    reviewGame.Title,
                    createCommentDto.ReviewGameId
                );
            }
            catch (Exception notificationEx)
            {
                logger.LogError(notificationEx,
                    "Errore nella creazione della notifica per commento recensione");
            }

            return mapper.Map<ReviewCommentDto>(savedComment);
        }

        /// <summary>
        /// Elimina un commento a una recensione
        /// </summary>
        public async Task<bool> DeleteReviewCommentAsync(int commentId, int userId)
        {
            var comment = await context.ReviewComments
                .FirstOrDefaultAsync(rc => rc.Id == commentId);

            if (comment == null)
                throw new KeyNotFoundException("Commento non trovato.");

            if (comment.AuthorId != userId)
                throw new UnauthorizedAccessException("Non sei autorizzato ad eliminare questo commento.");

            context.ReviewComments.Remove(comment);
            await context.SaveChangesAsync();

            return true;
        }

        /// <summary>
        /// Ottiene tutti i commenti per un'attività
        /// </summary>
        public async Task<List<ActivityCommentDto>> GetActivityCommentsAsync(int activityId)
        {
            var comments = await context.ActivityComments
                .Include(ac => ac.Author)
                .Where(ac => ac.ActivityId == activityId)
                .OrderBy(ac => ac.Date)
                .ToListAsync();

            return mapper.Map<List<ActivityCommentDto>>(comments);
        }

        /// <summary>
        /// Aggiunge un commento a un'attività
        /// </summary>
        public async Task<ActivityCommentDto?> AddActivityCommentAsync(CreateActivityCommentDto createCommentDto,
            int authorId)
        {
            var activity = await context.Activities
                .Include(a => a.Game)
                .FirstOrDefaultAsync(a => a.Id == createCommentDto.ActivityId);

            if (activity == null)
                throw new KeyNotFoundException("Attività non trovata.");

            var newComment = new ActivityComment
            {
                Text = createCommentDto.Text,
                Date = DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffZ"),
                ActivityId = createCommentDto.ActivityId,
                AuthorId = authorId
            };

            context.ActivityComments.Add(newComment);
            await context.SaveChangesAsync();

            var savedComment = await context.ActivityComments
                .Include(ac => ac.Author)
                .FirstOrDefaultAsync(ac => ac.Id == newComment.Id);

            if (savedComment == null) return null;

            // Crea notifica per il proprietario dell'attività (solo per attività di tipo "Rated")
            if (activity is { Game: not null, Type: ActivityType.Rated } &&
                activity.Game.UserId != authorId)
            {
                try
                {
                    await notificationService.CreateReviewCommentNotificationAsync(
                        activity.Game.UserId,
                        authorId,
                        savedComment.Author?.UserName ?? "Utente sconosciuto",
                        activity.GameTitle,
                        createCommentDto.ActivityId
                    );
                }
                catch (Exception notificationEx)
                {
                    logger.LogError(notificationEx,
                        "Errore nella creazione della notifica per commento all'attività");
                }
            }

            return mapper.Map<ActivityCommentDto>(savedComment);
        }

        /// <summary>
        /// Elimina un commento a un'attività
        /// </summary>
        public async Task<bool> DeleteActivityCommentAsync(int commentId, int userId)
        {
            var comment = await context.ActivityComments
                .FirstOrDefaultAsync(ac => ac.Id == commentId);

            if (comment == null)
                throw new KeyNotFoundException("Commento non trovato.");

            if (comment.AuthorId != userId)
                throw new UnauthorizedAccessException("Non sei autorizzato ad eliminare questo commento.");

            context.ActivityComments.Remove(comment);
            await context.SaveChangesAsync();

            return true;
        }

        private static AspectStatsDto CalculateAspectStats(IEnumerable<decimal> values)
        {
            var enumerable = values as decimal[] ?? values.ToArray();
            if (!(enumerable.Length > 0))
            {
                return new AspectStatsDto();
            }

            var valuesList = enumerable.ToList();

            return new AspectStatsDto
            {
                Average = Math.Round(valuesList.Average(), 2),
                Min = valuesList.Min(),
                Max = valuesList.Max(),
                Distribution = valuesList
                    .GroupBy(v => (int)Math.Round(v))
                    .ToDictionary(g => g.Key, g => g.Count())
            };
        }
    }
}

File: frontend/src/store/services/communityService.ts
import axios from 'axios';
import { 
  CommunityStatsDto, 
  CommunityReviewDto, 
  PaginatedReviewsDto, 
  ReviewStatsDto,
  CommunityRatingsResponse,
  CommunityRatingsWithCountResponse,
  ReviewCommentDto,
  CreateReviewCommentDto
} from '../../types/community';
import { ActivityComment, CreateActivityCommentDto } from '../../types/activity';
import { getToken } from '../../utils/getToken';
import { API_CONFIG, buildApiUrl } from '../../config/api';

const API_BASE_URL = buildApiUrl(API_CONFIG.ENDPOINTS.COMMUNITY);

// Crea istanza axios per le notifiche
const communityApi = axios.create({
  baseURL: API_BASE_URL
});

// Istanza axios che aggiunge il token JWT se presente
communityApi.interceptors.request.use((config) => {
  const token = getToken();
  if (token) {
    config.headers = config.headers || {};
    config.headers['Authorization'] = `Bearer ${token}`;
  }
  return config;
});

export class CommunityService {
  
  /**
   * Ottiene le statistiche della community per un gioco
   */
  static async getCommunityStats(gameTitle: string): Promise<CommunityStatsDto> {
    try {
        const response = await communityApi.get<CommunityStatsDto>(
            `/stats/${encodeURIComponent(gameTitle)}`
        );
      return response.data;
    } catch (error) {
      console.error('Errore nel recupero delle statistiche:', error);
      return {
        totalPlayers: 0,
        averageRating: 0,
        totalReviews: 0,
        averagePlaytime: 0,
        completionRate: 0,
        currentlyPlaying: 0
      };
    }
  }

  /**
   * Ottiene il rating medio della community per un gioco
   */
  static async getCommunityRating(gameTitle: string): Promise<number> {
    try {
      const response = await communityApi.get<number>(
        `/rating/${encodeURIComponent(gameTitle)}`
      );
      return response.data;
    } catch (error) {
      console.error('Errore nel recupero del rating:', error);
      return 0;
    }
  }

  /**
   * Ottiene i rating della community per una lista di giochi
   */
  static async getCommunityRatings(gameTitles: string[]): Promise<CommunityRatingsResponse> {
    try {
      const response = await communityApi.post<CommunityRatingsResponse>(
        '/ratings',
        gameTitles
      );
      return response.data;
    } catch (error) {
      console.error('Errore nel recupero dei rating:', error);
      return {};
    }
  }

  /**
   * Ottiene i rating della community con numero di recensioni per una lista di giochi
   */
  static async getCommunityRatingsWithCount(gameTitles: string[]): Promise<CommunityRatingsWithCountResponse> {
    try {
      const response = await communityApi.post<CommunityRatingsWithCountResponse>(
        '/ratings-with-count',
        gameTitles
      );
      return response.data;
    } catch (error) {
      console.error('Errore nel recupero dei rating con conteggio:', error);
      return {};
    }
  }

  /**
   * Ottiene le recensioni per un gioco con paginazione
   */
  static async getReviews(
    gameTitle: string, 
    page: number = 1, 
    pageSize: number = 10
  ): Promise<PaginatedReviewsDto> {
    try {
      const response = await communityApi.get<PaginatedReviewsDto>(
        `/reviews/${encodeURIComponent(gameTitle)}`,
        {
          params: { page, pageSize }
        }
      );
      return response.data;
    } catch (error) {
      console.error('Errore nel recupero delle recensioni:', error);
      return {
        reviews: [],
        totalCount: 0,
        page: 1,
        pageSize: 10,
        totalPages: 0
      };
    }
  }
  /**
   * Ottiene le recensioni pubbliche per un gioco con paginazione (senza la recensione dell'utente corrente)
   */
  static async getPublicReviews(
    gameTitle: string, 
    page: number = 1, 
    pageSize: number = 10
  ): Promise<PaginatedReviewsDto> {
    try {
      const response = await communityApi.get<PaginatedReviewsDto>(
        `/reviews/${encodeURIComponent(gameTitle)}`,
        {
          params: { page, pageSize }
        }
      );
      return response.data;
    } catch (error) {
      console.error('Errore nel recupero delle recensioni pubbliche:', error);
      return {
        reviews: [],
        totalCount: 0,
        page: 1,
        pageSize: 10,
        totalPages: 0
      };
    }
  }

  /**
   * Ottiene le statistiche dettagliate delle recensioni per un gioco
   */
  static async getReviewStats(gameTitle: string): Promise<ReviewStatsDto> {
    try {
      const response = await communityApi.get<ReviewStatsDto>(
        `/review-stats/${encodeURIComponent(gameTitle)}`
      );
      return response.data;
    } catch (error) {
      console.error('Errore nel recupero delle statistiche recensioni:', error);
      return {
        gameTitle,
        totalReviews: 0,
        averageGameplay: 0,
        averageGraphics: 0,
        averageStory: 0,
        averageSound: 0,
        overallAverageRating: 0,
        ratingDistribution: {},
        gameplayStats: { average: 0, min: 0, max: 0, distribution: {} },
        graphicsStats: { average: 0, min: 0, max: 0, distribution: {} },
        storyStats: { average: 0, min: 0, max: 0, distribution: {} },
        soundStats: { average: 0, min: 0, max: 0, distribution: {} }
      };
    }
  }

  /**
   * Ottiene i commenti per una specifica recensione
   */
  static async getReviewComments(reviewGameId: number): Promise<ReviewCommentDto[]> {
    try {
      const response = await communityApi.get<ReviewCommentDto[]>(
        `/review-comments/${reviewGameId}`
      );
      return response.data;
    } catch (error) {
      console.error('Errore nel recupero dei commenti:', error);
      return [];
    }
  }

  /**
   * Aggiunge un commento a una recensione
   */
  static async addReviewComment(createCommentDto: CreateReviewCommentDto): Promise<ReviewCommentDto | null> {
    try {
      const response = await communityApi.post<ReviewCommentDto>(
        '/review-comments',
        createCommentDto
      );
      return response.data;
    } catch (error) {
      console.error('Errore nell\'aggiunta del commento:', error);
      return null;
    }
  }
  /**
   * Elimina un commento a una recensione
   */
  static async deleteReviewComment(commentId: number): Promise<boolean> {
    try {
      await communityApi.delete(`/review-comments/${commentId}`);
      return true;
    } catch (error) {
      console.error('Errore nell\'eliminazione del commento:', error);
      return false;
    }
  }

  // ============ COMMENTI ALLE ATTIVITÀ ============

  /**
   * Ottiene tutti i commenti per un'attività
   */
  static async getActivityComments(activityId: number): Promise<ActivityComment[]> {
    try {
      const response = await communityApi.get<ActivityComment[]>(
        `/activity-comments/${activityId}`
      );
      return response.data;
    } catch (error) {
      console.error('Errore nel recupero dei commenti dell\'attività:', error);
      return [];
    }
  }

  /**
   * Aggiunge un commento a un'attività
   */
  static async addActivityComment(createCommentDto: CreateActivityCommentDto): Promise<ActivityComment | null> {
    try {
      const response = await communityApi.post<ActivityComment>(
        '/activity-comments',
        createCommentDto
      );
      return response.data;
    } catch (error) {
      console.error('Errore nell\'aggiunta del commento all\'attività:', error);
      return null;
    }
  }

  /**
   * Elimina un commento a un'attività
   */
  static async deleteActivityComment(commentId: number): Promise<boolean> {
    try {
      await communityApi.delete(`/activity-comments/${commentId}`);
      return true;
    } catch (error) {
      console.error('Errore nell\'eliminazione del commento dell\'attività:', error);
      return false;
    }
  }

  /**
   * Aggiunge una nuova recensione
   */
  static async addReview(reviewData: any): Promise<boolean> {
    try {
      await communityApi.post('/reviews', reviewData);
      return true;
    } catch (error) {
      console.error('Errore nell\'aggiunta della recensione:', error);
      return false;
    }
  }

}
